### PR TITLE
fix: preview on floating window

### LINKED ISSUE #1643: Error when calling preview when on floating mode
### Description

Clicking `<Tab>` to preview a file when floating mode is enabled results to `Invalid window id: 1002` error 

image

The issue occurs in this line https://github.com/nvim-tree/nvim-tree.lua/blob/master/lua/nvim-tree/view.lua#L339 since nvim-tree on floating mode is auto_closed by default, the `winnr` is always invalid since the window is already gone, I suggest using `a.nvim_win_is_valid(wnr)` before trying to focus the window.

I'm not sure what is the expected behavior here, should it reopens the floating window again when previewing a file 

### Neovim version

```text
NVIM v0.9.0-dev-1214-ge54541f7f-dirty
Build type: Release
LuaJIT 2.1.0-beta3
```

### Operating system and version

MacOS 12.6

### nvim-tree version

main - b01e7be

### Minimal config

```lua
vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvt-min/site]])
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
	require("packer").startup({
		{
			"wbthomason/packer.nvim",
			"kyazdani42/nvim-tree.lua",
			"kyazdani42/nvim-web-devicons",
			-- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
		},
		config = {
			package_root = package_root,
			compile_path = install_path .. "/plugin/packer_compiled.lua",
			display = { non_interactive = true },
		},
	})
end
if vim.fn.isdirectory(install_path) == 0 then
	print("Installing nvim-tree and dependencies.")
	vim.fn.system({ "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path })
end
load_plugins()
require("packer").sync()
vim.cmd([[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]])
vim.opt.termguicolors = true
vim.opt.cursorline = true

-- MODIFY NVIM-TREE SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
_G.setup = function()
	require("nvim-tree").setup({
		view = {
			float = {
				enable = true,
			},
		},
	})
end

vim.api.nvim_set_keymap("n", "<Space>e", "<cmd>:NvimTreeToggle<CR>", {})
```

### Steps to reproduce

1- nvim -nu /tmp/nvt-min.lua
2- :NvimTreeToggle
3- press `<Tab>` on any file

### Expected behavior

_No response_

### Actual behavior

_No response_

### PR DESCRIPTION
**Pull request by Almo7aya**
fixes #1643

My changes prevent nvim-tree from using already closed window, usually when using floating window mode.

More details in the issue #1643

### CODE REVIEW

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Testing with nvt-min.lua from the issue resulted in an empty nvim-tree window. Upon `R` the contents were correctly repopulated.

[asciicast](https://asciinema.org/a/528519)

**Comment by Almo7aya** [comment]
Ummmm, after some digging I found out that I have 
```lua
update_focused_file = {
	enable = true,
},
```
in my config file, when removing it the issue you faced occurs, I'm checking...

You can try this config with my branch and `update_focused_file = true`

```lua
vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvt-min/site]])
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
	require("packer").startup({
		{
			"wbthomason/packer.nvim",
			{
				"almo7aya/nvim-tree.lua",
				branch = "fix/preview_action_on_floating_windows",
			},
			"kyazdani42/nvim-web-devicons",
			-- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
		},
		config = {
			package_root = package_root,
			compile_path = install_path .. "/plugin/packer_compiled.lua",
			display = { non_interactive = true },
		},
	})
end
if vim.fn.isdirectory(install_path) == 0 then
	print("Installing nvim-tree and dependencies.")
	vim.fn.system({ "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path })
end
load_plugins()
require("packer").sync()
vim.cmd([[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]])
vim.opt.termguicolors = true
vim.opt.cursorline = true

-- MODIFY NVIM-TREE SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
_G.setup = function()
	require("nvim-tree").setup({
		update_focused_file = {
			enable = true,
		},
		view = {
			float = {
				enable = true,
			},
		},
	})
end

vim.api.nvim_set_keymap("n", "<Space>e", "<cmd>:NvimTreeToggle<CR>", {})

```

**Comment by alex-courtis** [comment]
> You can try this config with my branch and `update_focused_file = true`

That works as expected.

**Comment by Almo7aya** [comment]
seems like 
```lua
  if opts.update_focused_file.enable then
    create_nvim_tree_autocmd("BufEnter", {
      callback = function()
        M.find_file(false)
      end,
    })
  end
```
has something todo with this, removing `M.find_file(false)` results in the same issue you mentioned

**Comment by alex-courtis** [comment]
Cursory investigation:

It seems that the tree window is being closed during the preview action: https://github.com/Almo7aya/nvim-tree.lua/blob/79f631bc1d52b387f4ae59fad1291d894afa97f5/lua/nvim-tree/actions/node/open-file.lua#L134

Rather than being reactive to the closed window in `focus`, we could keep the floating tree window open.

**Comment by Almo7aya** [comment]
After a deep debugging sessions, I discovered that we have an option called `view.float.quit_on_focus_loss` which is enabled by default to kill the window when losing focus 
```lua
  if opts.view.float.enable and opts.view.float.quit_on_focus_loss then
    create_nvim_tree_autocmd("WinLeave", { pattern = "NvimTree_*", callback = view.close })
  end
``` 
so ignoring this autocmd on preview fixes the issue

**Comment by alex-courtis** [comment]
Still blank after preview. The floating window is definitely still being closed. This can be validated via `:echo win_getid()` before and after the preview.

The next call to `set_current_win_no_autocmd` is the one actually causing the WinLeave event to be fired. This could be moved into an else, something like
```lua
  if mode == "preview" then
    -- ignore "WinLeave" autocmd on preview
    -- because the registered "WinLeave"
    -- will kill the floating window immediately
    set_current_win_no_autocmd(api.nvim_get_current_win(), "WinLeave")
  else
    set_current_win_no_autocmd(target_winid)
  end
```

Unfortunately there is another WinLeave event fired after `open_in_new_window`.

**Comment by Almo7aya** [comment]
I was calling `set_current_win_no_autocmd ` with the wrong window id, I adjust it now, I can't replicate the blank window after preview

**Comment by alex-courtis** [comment]
> I was calling `set_current_win_no_autocmd ` with the wrong window id, I adjust it now, I can't replicate the blank window after preview

Nice work. I'm out of time today, I'll take it for a drive tomorrow.

File: lua/nvim-tree/view.lua [hunk 1]
```diff
@@ -335,8 +335,10 @@ end
 
 function M.focus(winnr, open_if_closed)
   local wnr = winnr or M.get_winnr()
-
-  if a.nvim_win_get_tabpage(wnr or 0) ~= a.nvim_win_get_tabpage(0) then
+  if not a.nvim_win_is_valid(wnr) then
```
Comment (alex-courtis) [inline lua/nvim-tree/view.lua:338]: I don't think we need this any more.

I tested successfully without these changes.

File: lua/nvim-tree/actions/node/open-file.lua [hunk 2]
```diff
@@ -209,7 +209,15 @@ local function open_in_new_window(filename, mode, win_ids)
     cmd = string.format("edit %s", fname)
   end
 
-  set_current_win_no_autocmd(target_winid)
+  if mode == "preview" then
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/node/open-file.lua:212]: This is correct, however is only desirable for floating windows. There may be (future) WinLeave events in the non-floating case that we still wish to receive.

- [ ] Please change to `if mode == "preview" and view float is set then`

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Tested OK with open-file changes only:

- [ ] float enabled
- [ ] float disabled
- [ ] filter visible
- [ ] quit_on_focus_lost disabled

**Comment by alex-courtis** [comment]
Thank you so much for this fix!

**Pull request merged**